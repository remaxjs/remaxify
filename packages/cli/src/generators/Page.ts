import * as t from '@babel/types';
import { addNamed } from '@babel/helper-module-imports';
import generator from '@babel/generator';
import traverse from '@babel/traverse';
import prettier from 'prettier';
import * as sander from 'sander';
import * as path from 'path';
import { options } from '../options';
import { RUNTIME_PACKAGE } from '../constants';
import { File } from '@babel/types';
import { SourceType } from '../sourceType';

const REMAXIFY_PAGE = 'RemaxifyPage';
const Page = SourceType.Page;

export default function generate(ast: File | undefined, filename: string): OutputFile {
  if (!ast) {
    throw new Error('Page 文件无法分析 ast');
  }

  filename = path.join(options.rootDir, filename);

  traverse(ast, {
    CallExpression(nodePath) {
      const { node } = nodePath;

      if (!t.isIdentifier(node.callee)) {
        return;
      }

      if (node.callee.name !== Page) {
        return;
      }

      const binding = nodePath.scope.getBinding(Page);

      if (binding) {
        return;
      }

      const appId = addNamed(nodePath, Page, RUNTIME_PACKAGE, {
        nameHint: REMAXIFY_PAGE,
      });
      const renderId = addNamed(nodePath, 'render', './render', {
        nameHint: 'render',
      });
      nodePath.replaceWith(t.callExpression(appId, [...node.arguments, renderId]));
      nodePath.stop();
    },
  });

  const code = generator(ast).code;
  const content = prettier.format(code, {
    parser: 'babel',
  });

  if (!options.dry) {
    sander.writeFile(options.output, filename, content);
  }

  return {
    filename,
    content,
  };
}
