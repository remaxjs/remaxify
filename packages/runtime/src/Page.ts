import * as React from 'react';

export default function Page(
  pageConfig: { [key: string]: any },
  render: () => React.ReactNode
): React.ComponentType {
  return class Page extends React.Component {
    constructor(props: any) {
      super(props);

      Object.keys(pageConfig).forEach(key => {
        if (typeof pageConfig[key] === 'function') {
          (this as any)[key] = pageConfig[key].bind(this);
          return;
        }

        (this as any)[key] = pageConfig[key];
      });

      this.state = pageConfig.data;
    }

    public data = this.state;

    public setData(payload: any, callback?: () => void) {
      return this.setState(payload, callback);
    }

    public render() {
      return render.bind(this)();
    }
  };
}
