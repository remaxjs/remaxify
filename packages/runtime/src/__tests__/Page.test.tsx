import * as React from 'react';
import TestRenderer from 'react-test-renderer';
import Page from '../Page';

describe('Page', () => {
  it('works', () => {
    const mockShow = jest.fn();
    const config: any = {
      data: {
        propA: 'A',
        propB: 'B',
      },

      onShow() {
        mockShow();
      },

      onTap() {
        this.setData({
          propA: 'AA',
          propB: 'BB',
        });
      },
    };
    const View = function ({ onTap, children }: any) {
      return React.createElement('view', { onTap, children });
    };

    function render(this: any) {
      return (
        <View onTap={this.onTap}>
          {this.state.propA}
          {this.state.propB}
        </View>
      );
    }
    const PageComponent = Page(config, render);

    const testRenderer = TestRenderer.create(<PageComponent />);

    expect(testRenderer.toJSON()).toMatchSnapshot();

    testRenderer.root.findByType('view').props.onTap();

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
