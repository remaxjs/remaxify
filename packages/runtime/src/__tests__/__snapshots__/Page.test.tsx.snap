// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page works 1`] = `
<view
  onTap={[Function]}
>
  A
  B
</view>
`;

exports[`Page works 2`] = `
<view
  onTap={[Function]}
>
  AA
  BB
</view>
`;
